File: src/solvers/s_lgmres_e.m
function [x, flag, relresvec, kdvec, time] = ...
    s_lgmres_e(A, b, mInitial, mMinMax, mStep, tol, maxit, xInitial, alphaPD, ...
             eps0, l, d, varargin)
    % Adaptive SLGMRES-E with Proportional-Derivative Controller
    %
    %   Description:
    %   ------------
    %
    %   Adaptive SLGMRES-E a hybrid method for improving GMRES(m) by
    %   restraining the stagnation and slowdown convergence. The idea behind
    %   the variation of the structure of the GMRES(m) relies on the fact that
    %   once either the slowdown of convergence or stagnation is detected,
    %   a rule of switching between two strategies chooses the one most
    %   appropriate for facing the identified convergence problem. For this
    %   setup. two monotonically decreasing strategies, obtained from the
    %   literature, are combined: the LGMRES and the GMRES-E, this results
    %   in the switching method denoted as SLGMRES-E. In some cases, a
    %   proportional derivative rule is introduced to modify the restart
    %   parameter 'm', which gives place to this Adaptive SLGMRES-E
    %   algorithm.
    %
    %   Signature:
    %   ----------
    %
    %   [x, flag, relresvec, kdvec, time] = s_lgmres_e(A, b, ...
    %       mInitial, mMinMax, mStep, tol, maxit, xInitial, alphaPD)
    %
    %
    %   Input Parameters:
    %   -----------------
    %
    %   A:          n-by-n matrix
    %               Left-hand side of the linear system Ax = b.
    %
    %   b:          n-by-1 vector
    %               Right-hand side of the linear system Ax = b.
    %
    %   mInitial:   int, optional
    %               Initial restart parameter (similar to 'restart' in MATLAB).
    %               If empty, not given, or equal to n, then mInitial is not
    %               used and the full unrestarted gmres algorithm with
    %               maxit = min(n, 10) is employed. Note that we require
    %               1 <= mInitial <= n.
    %
    %   mMinMax:    2-by-1 vector, optional
    %               Minimum and maximum values of the restart paramter m.
    %               Default is [1; n]. We require
    %               1 <= mMinMax(1) < mMinMax(2) <= n. Note that for large
    %               matrices (e.g., > 600), the default value mMinMax(2) = n
    %               might require large computational resources.
    %
    %   mStep:      int, optional
    %               Step size for increasing the restart parameter m when
    %               m < mMinMax(1). Default is 1 if n <= 10 and 3 otherwise.
    %
    %   tol:        float, optional
    %               Tolerance error threshold for the relative residual norm.
    %               Default is 1e-6.
    %
    %   maxit:      int, optional
    %               Maximum number of outer iterations.
    %
    %   xInitial:   n-by-1 vector, optional
    %               Vector of initial guess. Default is zeros(n, 1).
    %
    %   alphaPD:    2-by-1 vector, optional
    %               Proportional and derivative coefficients from the
    %               proportional-derivative controller. Default is [-3, 5],
    %               following Ref. [2].
    %
    %   eps0:       float
    %               Stagnation threshold parameter, as defined in [1].
    %               Slow convergence occurs when ||y|| < eps0,
    %               and approximate error vectors don't have enough
    %               impact.
    %
    %   Output parameters:
    %   ------------------
    %
    %   x:          n-by-1 vector
    %               Approximate solution of the linear system.
    %
    %   flag:       boolean
    %               1 if the algorithm has converged, 0 otherwise.
    %
    %   relressvec: (1 up to maxit)-by-1 vector
    %               Vector of relative residual norms of every outer iteration
    %               (cycles). The last relative residual norm is simply given
    %               by relresvec(end).
    %
    %   kdvec:      (1 up to maxit)-by-1 vector
    %               Vector of restart parameter values. In case the unrestarted
    %               algorithm is invoked, kdvec = NaN.
    %
    %   time:       scalar
    %               Computational time in seconds.
    %
    %   References:
    %   -----------
    %
    %   [1] Cabral, J. C., Schaerer, C. E., & Bhaya, A. (2020). Improving
    %   GMRES(m) using an adaptive switching controller. Numerical Linear
    %   Algebra with Applications, 27(5), e2305.
    %
    %   Copyright:
    %   ----------
    %
    %   This file is part of the KrySBAS MATLAB Toolbox.
    %
    %   Copyright 2023 CC&MA - NIDTec - FP - UNA
    %
    %   KrySBAS is free software: you can redistribute it and/or modify it under
    %   the terms of the GNU General Public License as published by the Free
    %   Software Foundation, either version 3 of the License, or (at your
    %   option) any later version.
    %
    %   KrySBAS is distributed in the hope that it will be useful, but WITHOUT
    %   ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
    %   FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
    %   for more details.
    %
    %   You should have received a copy of the GNU General Public License along
    %   with this file.  If not, see <http://www.gnu.org/licenses/>.
    %

    % ----> Sanity check on the number of input parameters
    if nargin < 2
        error("Too few input parameters. Expected at least A and b.");
    elseif nargin > 9
        error("Too many input parameters.");
    end

    % ----> Sanity checks on matrix A
    % Check whether A is non-empty
    if isempty(A)
        error("Matrix A cannot be empty.");
    end

    % Check whether A is square
    [rowsA, colsA] = size(A);
    if rowsA ~= colsA
        error("Matrix A must be square.");
    end

    n = rowsA;
    clear rowsA colsA;

    % ----> Sanity checks on vector b
    % Check whether b is non-empty
    if isempty(b)
        error("Vector b cannot be empty.");
    end

    % Check whether b is a column vector
    [rowsb, colsb] = size(b);
    if colsb ~= 1
        error("Vector b must be a column vector.");
    end

    % Check whether b has the same number of rows as b
    if rowsb ~= n
        error("Dimension mismatch between matrix A and vector b.");
    end

    clear rowsb colsb;

    % ----> Default values and sanity checks for parameter mInitial
    % When the restart parameter is not specified explicitly as an input
    % argument, pd_gmres() behaves identically to the unrestarted gmres()
    % with a fixed number of iterations given by min(n, 10).
    %
    % There are three possibilities when this might happen:
    %   (1) If only two arguments are given i.e., A and b.
    %   (2) If mInitial equals the dimension of A i.e., mInitial = n.
    %   (3) If an empty matrix is given as mInitial, i.e., mInitial = [].
    if (nargin < 3) || isempty(mInitial) || (mInitial == n)
        restarted = false;
    else
        restarted = true;
    end

    % If the restarted version of pd_gmres will be used, then the value of
    % mInitial must be bounded between 1 and n.
    if restarted && (mInitial < 1 || mInitial > n)
        error("mInitial must satisfy: 1 <= mInitial <= n.");
    end

    % ----> Default values and sanity checks for mMinMax
    if (nargin < 4) || isempty(mMinMax)
        mMinMax = [1; n];
    else
        if ~restarted
            warning("mMinMax was given but will not be used.");
        end
    end

    if restarted && ...
            ((mMinMax(1) < 1) || ...
             (mMinMax(2) > n) || ...
             (mMinMax(2) <= mMinMax(1)))
        error("mMinMax must satisfy: 1 <= mMinMax(1) < mMinMax(2) <= n.");
    end

    if restarted && ((mMinMax(1) > mInitial) || (mMinMax(2) < mInitial))
        error("mMinMax must satisfy: mMinMax(1) <= mInitial <= mMinMax(2).");
    end

    mMin = mMinMax(1);
    mMax = mMinMax(2);

    % ----> Default values and sanity checks for mStep
    if (nargin < 5) || isempty(mStep)
        mStep = 1;
    else
        if ~restarted
            warning("mStep was given but will not be used.");
        end
    end

    if (mStep < 1) || (mStep > n - 1)
        error("mStep must satisfy: 0 < mStep < n.");
    end

    % ----> Default value and sanity checks for tol
    if (nargin < 6) || isempty(tol)
        tol = 1e-6;
    end

    if tol < eps
        warning("Tolerance is too small and it will be changed to eps.");
        tol = eps;
    elseif tol >= 1
        warning("Tolerance is too large and it will be changed to 1-eps.");
        tol = 1 - eps;
    end

    % ----> Default value for maxit
    if (nargin < 7) || isempty(maxit)
        if restarted
            maxit = min(ceil(n / mInitial), 10);
        else
            maxit = min(n, 10);
        end
    end

    % ----> Default value and sanity checks for initial guess xInitial
    if (nargin < 8) || isempty(xInitial)
        xInitial = zeros(n, 1);
    end

    % ----> Default value for eps0
    if (nargin < 9) || isempty(eps0)
        eps0 = 0.01;
    end

    % Check whether xInitial is a column vector
    [rowsxInitial, colsxInitial] = size(xInitial);
    if colsxInitial ~= 1
        error("Initial guess xInitial is not a column vector.");
    end

    % Check whether x0 has the right dimension
    if rowsxInitial ~= n
        msg = "Dimension mismatch between matrix A and initial guess xInitial.";
        error(msg);
    end

    clear rowsxInitial colsxInitial;

    % ----> Default value for alphaPD
    % The default values for these parameteres were taken from page 217 of [2]
    if (nargin < 9) || isempty(alphaPD)
        alphaPD = [-3; 5];
    end

    alphaP = alphaPD(1);
    alphaD = alphaPD(2);

    % ---> S-LGMRES-E Algorithm starts here
    % First outer iteration consists of a single GMRES(m) iteration, plus a
    % switching strategy:
    %
    % * It it presents an appropriate good convergence, computing
    %   approximation error vectors is enough, as in the LGMRES mehtod
    % * If poor convergence is rather detected, then the Harmonic Ritz 
    %   vectors are computed as in the GMRES-E method, plus an update of the
    %   restart parameter 'm' via a proportional-derivative (PD) rule
    
    flag = 0;
    flag_stagnation = 0;
    restart = 1;
    r0 = b - A * xInitial;
    res(1, :) = norm(r0);
    relresvec(1, :) = (norm(r0) / res(1, 1));
    iter(1, :) = restart;
    beta = norm(r0);
    v1 = r0 / beta;
    kdvec(1, 1) = mInitial;

    tic();  % start measuring CPU time

    % Modified Gram-Schmidt Arnoldi iteration
    % This is the first run. Since we don't have
    % harmonic Ritz vectors yet, we use GMRES(m).
    s = mInitial;
    [H, V, sUpdated] = modified_gram_schmidt_arnoldi(A, v1, s);
    
    % Plane rotations
    [HUpTri, g] = plane_rotations(H, beta);

    % Solve the least-squares problem
    s = sUpdated;
    Rs = HUpTri(1:s, 1:s);
    gs = g(1:s);
    minimizer = Rs \ gs;
    zCurrentCycle = V * minimizer;
    xm = xInitial + zCurrentCycle;

    % Update residual norm, iterations, and relative residual vector
    res(restart + 1, :) = abs(g(s + 1, 1));
    iter(restart + 1, :) = restart + 1;
    relresvec(restart + 1, :) = res(restart + 1, :) / res(1, 1);
    
    % Check convergence
    if relresvec(restart + 1, :) < tol
        % We reached convergence.
        flag = 1;
        x = xm;
        time = toc();
        kdvec = s .* ones(length(relresvec), 1);
        return
    else
        % We have not reached convergence.
        % Now, we check for stagnation of the 'minimizer' vector
        if norm(minimizer) < eps0
            flag_stagnation = 1;
            Fold = H(1:s, 1:s)';
            G = Rs' * Rs;
            dy = harmonic_ritz_vectors(Fold, G, d, V, eigstol);

            % compute pdrule
        else
            % save zCurrentcycle
        end

        Fold = H(1:s, 1:s)';
        G = Rs' * Rs;
        dy = harmonic_ritz_vectors(Fold, G, d, V, eigstol);

        % Control block for 'm'
        if iter(size(iter, 1), :) ~= 1
            [miter] = pd_rule(m, n, mInitial, mMin, mMax, ...
                              mStep, res, iter(size(iter, 1), :), ...
                              alphaP, alphaD);
            m = miter(1, 1);
            mInitial = miter(1, 2);
        else
            m = mInitial;
        end
        kdvec(iter(size(iter, 1), :) + 1, 1) = m;

        % Update and restart.
        restart = restart + 1;
    end

    % Empty matrix E for next outer iteration
    E = zeros(s, d);
    
    % ---> S-LGMRES-E Algorithm for restart > 1
    while flag == 0 && restart <= maxit

        % Compute normalized residual vector
        r = b - A * xm;
        beta = norm(r);
        v1 = r / beta;

        % Now we check for the stagnation (or not) of the previous cycle
        % * If flag_stagnation == 0, then we generate the Arnoldi basis using
        %   the error approximation vectors,
        % * If flag_stagnation == 1, then we generate the Arnoldi basis using
        %   the harmonic Ritz vectors.
        if flag_stagnation == 0
            %s = mj + l;
            [H, V, s] = ...
                augmented_gram_schmidt_arnoldi(A, v1, m, zMat(:, 1:min(nevec, l)));
        else
            %s = mj + d;
            [H, V, s] = ...
                augmented_gram_schmidt_arnoldi(A, v1, m, fliplr(dy(:, 1:d)));
        end

        % To be continued...

    end
    
    time = toc();

end

File: julia/src/solvers/slgmres_e.jl
"""
    slgmres_e(A, b; m=0, l=-1, d=-1, epsilon_threshold=0.01, tol=1e-6, maxit=0, x_initial=[], eigstol=1e-6)

Restarted GMRES that switches, cycle by cycle, between two augmentation
strategies based on an observed convergence-slowdown signal (SLGMRES-E(*m*,
*l*, *d*)).

By default each cycle behaves like LGMRES(*m*, *l*): the restart subspace is
augmented with up to *l* error approximation vectors from prior cycles. When
slowdown is detected -- the relative residual ratio between two consecutive
cycles exceeds `1 - epsilon_threshold` -- it switches to GMRES-E(*m*, *d*)
for one or more cycles: the subspace is instead augmented with *d* harmonic
Ritz vectors computed from the cycle that triggered the detection. It
switches back to LGMRES-style augmentation as soon as the ratio improves
again.

The slowdown signal, `epsilon` in `‖r_m^(j)‖ / ‖r_m^(j-1)‖ = 1 - epsilon`, is
read directly off the Givens-rotated residual already computed for the
least-squares solve every cycle -- no extra cost. Under this switching rule
the residual norm is provably non-increasing cycle to cycle (Theorem 2 of
[1]).

# Arguments
- `A`: square coefficient matrix (sparse or dense, `n×n`)
- `b::AbstractVector`: right-hand side vector of length `n`
- `m::Int=0`: restart parameter, fixed across all cycles; defaults to
  `min(n, 10)`. Setting `m == n` dispatches to full unrestarted GMRES.
- `l::Int=-1`: number of error approximation vectors to append during
  LGMRES-style cycles; defaults to `3`. Must satisfy `l > 0`.
- `d::Int=-1`: number of harmonic Ritz vectors to append during GMRES-E
  -style cycles; defaults to `min(m, 3)`. Must satisfy `d > 0`.
- `epsilon_threshold::Real=0.01`: slowdown threshold. A cycle is classified
  as stagnating when `‖r^(j)‖ / ‖r^(j-1)‖ >= 1 - epsilon_threshold`. Default
  matches the numerical experiments of [1].
- `tol::Real=1e-6`: relative residual tolerance for convergence
- `maxit::Int=0`: maximum number of restart cycles; defaults to `min(n, 10)`
- `x_initial::AbstractVector=[]`: initial guess; defaults to the zero vector
- `eigstol::Real=1e-6`: reserved for API compatibility (unused internally,
  since the underlying `harmonic_ritz_vectors` uses a dense eigensolver
  here, not an iterative one)

# Returns
- `x`: approximate solution vector
- `flag::Bool`: `true` if `relresvec[end] < tol` within `maxit` restarts
- `relresvec::Vector`: relative residual norm after each restart cycle
- `kdvec::Vector`: Krylov subspace dimension used at each cycle (`m` during
  cycle 1, `m+l` or `m+d` thereafter depending on which augmentation was
  active)
- `time::Float64`: elapsed wall-clock time in seconds

# References
Cabral, J. C., Schaerer, C. E., & Bhaya, A. (2020). Improving GMRES(m) using
an adaptive switching controller. *Numerical Linear Algebra with
Applications*, 27(5), e2305.
[doi:10.1002/nla.2305](https://doi.org/10.1002/nla.2305)
"""
function slgmres_e(
    A,
    b::AbstractVector;
    m::Int = 0,
    l::Int = -1,
    d::Int = -1,
    epsilon_threshold::Real = 0.01,
    tol::Real = 1e-6,
    maxit::Int = 0,
    x_initial::AbstractVector = Float64[],
    eigstol::Real = 1e-6,
)

    # Sanity checks
    if ndims(A) != 2 || size(A, 1) == 0
        throw(ArgumentError("Matrix A cannot be empty."))
    end
    if size(A, 1) != size(A, 2)
        throw(ArgumentError("Matrix A must be square."))
    end
    n = size(A, 1)

    if isempty(b)
        throw(ArgumentError("Vector b cannot be empty."))
    end
    if length(b) != n
        throw(ArgumentError("Dimension mismatch between matrix A and vector b."))
    end

    if m == 0
        m = min(n, 10)
    end

    if isempty(x_initial)
        x_initial = zeros(eltype(b), n)
    end
    if length(x_initial) != n
        throw(
            ArgumentError(
                "Dimension mismatch between matrix A and initial guess x_initial.",
            ),
        )
    end

    # Dispatch: m == n → full unrestarted GMRES
    if m == n
        t0 = time()
        x, stats = Krylov.gmres(A, b; memory = n)
        elapsed = time() - t0
        res0 = norm(b - A * x_initial)
        resf = norm(b - A * x)
        relresvec = [1.0, resf / res0]
        kdvec = fill(n, 2)
        return x, stats.solved, relresvec, kdvec, elapsed
    end

    if m > n
        throw(ArgumentError("m must satisfy: 1 <= m <= n."))
    end

    l = l == -1 ? 3 : l
    d = d == -1 ? min(m, 3) : d

    if l <= 0
        throw(ArgumentError("l must satisfy: l > 0."))
    end
    if d <= 0
        throw(ArgumentError("d must satisfy: d > 0."))
    end

    if epsilon_threshold <= 0 || epsilon_threshold >= 1
        throw(ArgumentError("epsilon_threshold must satisfy: 0 < epsilon_threshold < 1."))
    end

    eps_val = eps(Float64)
    if tol < eps_val
        @warn "Tolerance is too small; changed to eps."
        tol = eps_val
    elseif tol >= 1
        @warn "Tolerance is too large; changed to 1 - eps."
        tol = 1 - eps_val
    end

    if maxit == 0
        maxit = min(n, 10)
    end

    # --- SLGMRES-E algorithm ---
    T = eltype(b)
    norm_y = 1 - epsilon_threshold

    x = copy(x_initial)
    r0 = b - A * x
    res1 = norm(r0)

    relresvec = [1.0]
    kdvec = Int[]

    # Sliding window of LGMRES-style error approximation vectors, newest
    # vector last. n_z tracks how many columns are actually populated
    # (independent of how many cycles have elapsed, since GMRES-E-style
    # cycles do not touch this window).
    z_mat = zeros(T, n, l)
    n_z = 0

    t0 = time()

    # -------------------------------------------------------------------
    # Cycle 1: plain GMRES(m). Neither augmentation strategy has any
    # history to draw on yet.
    # -------------------------------------------------------------------
    v1 = r0 / res1
    H, V, s = modified_gram_schmidt_arnoldi(A, v1, m)
    h_up_tri, g = plane_rotations(H, res1)

    rs = h_up_tri[1:s, 1:s]
    gs = g[1:s]
    minimizer = rs \ gs
    z_cycle = V * minimizer
    x = x + z_cycle
    push!(relresvec, abs(g[s+1]) / res1)
    push!(kdvec, s)

    if relresvec[end] < tol
        return x, true, relresvec, kdvec, time() - t0
    end

    # The correction from this cycle is always kept as the first error
    # approximation vector, regardless of what the next cycle turns out
    # to be.
    z_mat[:, 1] = z_cycle
    n_z = 1

    stagnating = relresvec[end] / relresvec[end-1] >= norm_y
    dy = zeros(T, n, 0)
    if stagnating
        # Cycle 1's basis is a plain (non-augmented) Arnoldi basis, so the
        # cheap H'-based formula for fold is exact here (equivalent to,
        # but cheaper than, W'*A'*W). Later cycles cannot use this
        # shortcut because V's augmented columns get overwritten with raw
        # (non-Arnoldi-consistent) vectors below -- see the main loop.
        fold = H[1:s, 1:s]'
        g_mat = rs' * rs
        dy = harmonic_ritz_vectors(fold, g_mat, d, V)
    end

    # -------------------------------------------------------------------
    # Main loop: cycles 2, 3, ...
    # -------------------------------------------------------------------
    flag = false
    while !flag && length(relresvec) - 1 < maxit

        r = b - A * x
        beta = norm(r)
        v1 = r / beta

        local s_cyc
        if !stagnating
            # --- LGMRES-style cycle ---
            # Note the reverse() placement here is the opposite of the
            # GMRES-E-style branch below: matching lgmres.jl's own
            # convention (not gmres_e.jl's), the newest error vector goes
            # into the first augmentation slot V[:, m+1]. processing_order
            # is that same slot order (newest first); augmented_gram_schmidt
            # _arnoldi processes it in this order and, unlike the MATLAB
            # utility, may drop trailing (nearly-dependent) columns via its
            # own near-breakdown guard, so the V overwrite below must not
            # assume all l_use columns survived.
            l_use = min(n_z, l)
            processing_order = reverse(z_mat[:, 1:l_use], dims = 2)
            H, V, s_cyc = augmented_gram_schmidt_arnoldi(A, v1, m, z_mat[:, 1:l_use])
            h_up_tri, g = plane_rotations(H, beta)
            rs = h_up_tri[1:s_cyc, 1:s_cyc]
            gs = g[1:s_cyc]
            minimizer = rs \ gs
            V[:, (m+1):s_cyc] = processing_order[:, 1:(s_cyc-m)]
        else
            # --- GMRES-E-style cycle ---
            processing_order = dy[:, 1:d]
            H, V, s_cyc =
                augmented_gram_schmidt_arnoldi(A, v1, m, reverse(processing_order, dims = 2))
            h_up_tri, g = plane_rotations(H, beta)
            rs = h_up_tri[1:s_cyc, 1:s_cyc]
            gs = g[1:s_cyc]
            minimizer = rs \ gs
            V[:, (m+1):s_cyc] = processing_order[:, 1:(s_cyc-m)]
        end
        s = s_cyc

        aux = V * minimizer
        x = x + aux
        push!(relresvec, abs(g[s+1]) / res1)
        push!(kdvec, s)

        if relresvec[end] < tol
            flag = true
            break
        end

        # ------------------------------------------------------------------
        # Decide the augmentation strategy for the NEXT cycle from the
        # ratio just observed. This mirrors [1] eq. (33)-(35): the ratio is
        # read off the residual we already computed above, at no extra
        # cost. The decision is independent of which strategy produced THIS
        # cycle's result -- a recovering GMRES-E cycle's correction is just
        # as valid an error-approximation vector as an LGMRES cycle's, and
        # a newly-stagnating LGMRES cycle's basis is just as valid a source
        # of harmonic Ritz vectors as a GMRES-E cycle's.
        # ------------------------------------------------------------------
        ratio = relresvec[end] / relresvec[end-1]

        if ratio >= norm_y
            stagnating = true
            W = V[:, 1:s]
            fold = W' * A' * W
            g_mat = rs' * rs
            dy = harmonic_ritz_vectors(fold, g_mat, d, V)
        else
            stagnating = false
            if n_z < l
                n_z += 1
                z_mat[:, n_z] = aux
            else
                z_mat[:, 1:(l-1)] = z_mat[:, 2:l]
                z_mat[:, l] = aux
            end
        end
    end

    return x, flag, relresvec, kdvec, time() - t0
end
